### PR TITLE
retry leader read when stale read encounters data not ready

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What is changed and how it works?

Issue Number: Ref #14553 

What's Changed:
When the local peer handling the stale read encounters a 'data is not ready' error, if the current local peer is a valid leader, attempt to directly fallback to leader snapshot read processing instead of returning an error directly to the KV client.
Save unnecessary handling of 'data not ready' errors and gRPC request processing, because the KV client will fallback to the leader peer for snapshot read upon receiving a 'data not ready' error too.

From the local test using command
```
mysqlslap --user=root --host=127.0.0.1 --port=4000 --create-schema=test --concurrency=10 --iterations=10000 --query="set tidb_read_staleness = '-5'; use test; select * from sbtest1 where id = 1"
```

Master:
image
image
image

This PR:
image
image
image
image

- There's no data is not ready error in the kv client, and kv request number is only half compared with the vanilla version. 
- Retry locally also saves resources.

Resource usage compare

image

### Check List 

Tests 

- Unit test

Side effects

### Release note 

```release-note
Retry leader read when stale read encounters data not ready.
```

### CODE REVIEW

File: components/raftstore/src/store/worker/read.rs
Comment (ekexium) [inline components/raftstore/src/store/worker/read.rs:1103]: Shall we also record the number of failed attempts of fallback leader read?

Comment (cfzjywxk) [reply components/raftstore/src/store/worker/read.rs:1105]: Seperate metrics are added.

**Review by zyguan — COMMENTED** [review]
Do we need to port the enhancement to raftstore-v2?

**Comment by cfzjywxk** [comment]
> Do we need to port the enhancement to raftstore-v2?

@zyguan Yes,  this PR is just the v1 version. We may need to adapt all the improvements to v2 later in v7.5 release.

**Review by zyguan — APPROVED** [review]

**Comment by ekexium** [comment]
Why does it cost 30% more CPU with this PR? My intuition is that same QPS should result in the equivalent level of CPU usage

**Comment by cfzjywxk** [comment]
> Why does it cost 30% more CPU with this PR? My intuition is that same QPS should result in an equivalent level of CPU usage(, if not less because of the reduction of retries).

@ekexium 
Yes, the functional tests above were conducted on a single machine. In theory, the overall overhead on the master branch at the same QPS should be higher. It might be an error caused by noise in my development environment. Strangely, I tried it three times on my development machine, and the results were similar. The CPU usage of the unified read pool would be higher, but the overall QPS difference is not significant. 
I might need to deploy and test it again on the testing cluster.

**Comment by ekexium** [comment]
OK. We can wait until we figure it out to merge this PR.

**Comment by cfzjywxk** [comment]
> OK. We can wait until we figure it out to merge this PR.

The test cluster has similar phenonmenon, the retry number is reduced but the async snapshot is slower than the vanillia version so is the read pool task handling. It's quite weird as there is just an extra `inspect` operations before retry local leader read which should not bring much overhead.

**Comment by cfzjywxk** [comment]
> > OK. We can wait until we figure it out to merge this PR.
> 
> The test cluster has similar phenonmenon, the retry number is reduced but the async snapshot is slower than the vanillia version so is the read pool task handling. It's quite weird as there is just an extra `inspect` operation before retry local leader read which should not bring much overhead.

I think maybe I made a mistke, I'm comparing the results from release and debug versions..

**Comment by cfzjywxk** [comment]
@ekexium I've updated the new test result in the description, PTAL

**Comment by cfzjywxk** [comment]
@you06 PTAL

**Review by ekexium — APPROVED** [review]

File: components/raftstore/src/store/worker/read.rs
Comment (you06) [inline components/raftstore/src/store/worker/read.rs:1088]: Isn't this check little redundant? I agree with more checks and the metric might counted if the check failed here.

Comment (cfzjywxk) [reply components/raftstore/src/store/worker/read.rs:1088]: The `inspector` function contains the whole read policy checks including `has_applied_to_current_term`. If the returned policy is `LocalRader` finally lease check would be done in the `try_local_leader_read` function.

Comment (you06) [reply components/raftstore/src/store/worker/read.rs:1088]: If `allow_fallback_leader_read` is false, it's also a failure of stale read fallback, what about increasing the `local_executed_stale_read_fallback_failure_requests` metric?

Comment (cfzjywxk) [reply components/raftstore/src/store/worker/read.rs:1088]: If allow_fallback_leader_read is false, the peer is usually a follower peer which is not expected to be retried locally. Adding 1 to the counter for this situation seems to have little significance? Or do we prefer to record cases where we believe the leader can retry but the actual retry fails, indirectly indicating lease instability?

Comment (you06) [reply components/raftstore/src/store/worker/read.rs:1088]: > If `allow_fallback_leader_read` is false, the peer is usually a follower peer which is not expected to be retried locally.
I get it.

**Review by you06 — APPROVED** [review]

**Pull request merged**
